drivers/include/hdc2010.h

```diff
@@ -0,0 +1,208 @@
+ */
+#ifndef CONFIG_HDC2010_I2C_ADDRESS
+#define CONFIG_HDC2010_I2C_ADDRESS           (0x40)
+#endif
+
+/**
+ * @brief   Typical conversion time needed to acquire new values [in us]
+ *
+ * @note    This time value is chosen twice as large as needed for two 14-bit
+ *          conversions (worst case) to allow for timer imprecision:
+ *          (convert temp + convert hum) * 2 -> (610us + 660us) * 2 := 2540us.
+ */
+#ifndef CONFIG_HDC2010_CONVERSION_TIME
+#define CONFIG_HDC2010_CONVERSION_TIME       (2540)
+#endif
+/** @} */
+
+/**
+ * @brief   HDC2010 specific return values
+ */
+enum {
+    HDC2010_OK     = 0,     /**< everything went as expected */
+    HDC2010_NODEV  = -1,    /**< no HDC2010 device found on the bus */
+    HDC2010_NOBUS  = -2,    /**< errors while initializing the I2C bus */
+    HDC2010_BUSERR = -3     /**< error during I2C communication */
+};
+
+/**
+ * @brief   Possible resolution values
+ */
+typedef enum {
+    HDC2010_9BIT  = (HDC2010_TRES9 | HDC2010_HRES9),    /**< 9-bit conversion */
+    HDC2010_11BIT = (HDC2010_TRES11 | HDC2010_HRES11),  /**< 11-bit conversion */
+    HDC2010_14BIT = (HDC2010_TRES14 | HDC2010_HRES14)   /**< 14-bit conversion */
+} hdc2010_res_t;
+
+/**
+ * @brief   Possible Auto Measurement Modes
+ */
+typedef enum {
+    HDC2010_TRIGGER    = 0,                     /**< Trigger based measurment */
+    HDC2010_1_SMPL_2M  = HDC2010_SAMPLE_2M,     /**< 1 sample every 2 minutes */
+    HDC2010_1_SMPL_1M  = HDC2010_SAMPLE_1M,     /**< 1 sample every minute */
+    HDC2010_1_SMPL_10S = HDC2010_SAMPLE_10S,    /**< 1 sample every 10 seconds */
+    HDC2010_1_SMPL_5S  = HDC2010_SAMPLE_5S,     /**< 1 sample every 5 seconds */
+    HDC2010_1_SMPL_1S  = HDC2010_SAMPLE_1HZ,    /**< 1 sample every second */
+    HDC2010_2_SMPL_1S  = HDC2010_SAMPLE_2HZ,    /**< 2 samples every second */
+    HDC2010_5_SMPL_1S  = HDC2010_SAMPLE_5HZ     /**< 5 sample every second */
+} hdc2010_amm_t;
+
+/**
+ * @brief   Possible measurement modes
+ */
+typedef enum {
+    HDC2010_MEAS_HUM_AND_TEMP = 0,          /**< Measure both humidity and temperature */
+    HDC2010_MEAS_TEMP = HDC2010_TEMP_MODE,  /**< Only measure temperature */
+} hdc2010_meas_mode_t;
+
+
+/**
+ * @brief   Callback function when interrupt happens
+ *
+ * This allows the user to define custom interrupt logic.
+ */
+ typedef void (*hdc2010_int_cb_t)(void *arg);
+
+/**
+ * @brief   Parameters needed for device initialization
+ */
+typedef struct {
+    i2c_t i2c;                  /**< bus the device is connected to */
+    uint8_t addr;               /**< address on that bus */
+    gpio_t irq_pin;             /**< IRQ pin to check data ready */
+    hdc2010_res_t res;          /**< resolution used for sampling temp and hum */
+    hdc2010_meas_mode_t mode;   /**< sensors to be used */
+    hdc2010_amm_t amm;          /**< Auto Measurement Mode */
+    hdc2010_int_cb_t user_cb;   /**< User defined interrupt callback */
```

Comment:
What would that be used for?
Can't think of a reason why a user would hook the data ready interrupt of the sensor.

Reply:
so actually the story here is the following - maybe i should rather use this usecase in the test?

when you are using the AMM mode (meaning you specify a predefined frequency of measurement of the values), you will get IRQ from the chip itself when the data is read; so the pseudo code would look something like this:

 1) init the dev with your own callback function, where in the callback function you do a `hdc2010_get_results(...)`
 2) call a `hdc2010_trigger_conversion` to start the AMM, i.e. start measuring

every time there's a valid measurement by the chip, it'll call `user_cb`, where you call `hdc2010_get_results` - not results as no point of checking anything like which mode etc - so that you read out the new measurements.

Reply:
> every time there's a valid measurement by the chip, it'll call `user_cb`, where you call `hdc2010_get_results`

But you can't call `hdc2010_get_results` inside the callback since I2C operations are not allowed in interrupt context.

Instead of having the API user deal with that, why not simply do

```C
while (measuring) {
    hdc2010_read(dev, &temp, &hum);
    …
}
```

since `hdc2010_read()` is blocking you will get the periodic measurement you want.
You can also add a check to the function so it will only start the measurement once if periodic measurement is configured.
(There is no function to stop periodic measurement?)